src/Columns/ColumnLowCardinality.cpp

```diff
@@ -426,29 +426,85 @@ void ColumnLowCardinality::getPermutation(IColumn::PermutationSortDirection dire
     getPermutationImpl(direction, stability, limit, nan_direction_hint, res);
 }
 
-void ColumnLowCardinality::updatePermutation(IColumn::PermutationSortDirection direction, IColumn::PermutationSortStability stability,
-                                        size_t limit, int nan_direction_hint, IColumn::Permutation & res, EqualRanges & equal_ranges) const
+namespace
 {
-    bool ascending = direction == IColumn::PermutationSortDirection::Ascending;
 
-    auto comparator = [this, ascending, stability, nan_direction_hint](size_t lhs, size_t rhs)
+/// Comparator for sorting LowCardinality column with the help of sorted dictionary.
+/// NOTE: Dictionary itself must be sorted in ASC or DESC order depending on the requested direction.
+template <typename IndexColumn, bool stable>
+struct LowCardinalityComparator
+{
+    const IndexColumn & real_indexes;                   /// Indexes column
+    const PaddedPODArray<UInt64> & position_by_index;   /// Maps original dictionary index to position in sorted dictionary
+
+    inline bool operator () (size_t lhs, size_t rhs) const
     {
-        int ret = getDictionary().compareAt(getIndexes().getUInt(lhs), getIndexes().getUInt(rhs), getDictionary(), nan_direction_hint);
-        if (unlikely(stability == IColumn::PermutationSortStability::Stable && ret == 0))
+        int ret;
+
+        const UInt64 lhs_index = real_indexes.getUInt(lhs);
+        const UInt64 rhs_index = real_indexes.getUInt(rhs);
+
+        if (lhs_index == rhs_index)
+            ret = 0;
+        else
+            ret = CompareHelper<UInt64>::compare(position_by_index[lhs_index], position_by_index[rhs_index], 0);
+
+        if (stable && ret == 0)
             return lhs < rhs;
 
-        if (ascending)
-            return ret < 0;
-        return ret > 0;
-    };
+        return ret < 0;
+    }
+};
 
-    auto equal_comparator = [this, nan_direction_hint](size_t lhs, size_t rhs)
+}
+
+template <typename IndexColumn>
+void ColumnLowCardinality::updatePermutationWithIndexType(
+    IColumn::PermutationSortDirection direction, IColumn::PermutationSortStability stability,
+    size_t limit, int nan_direction_hint, IColumn::Permutation & res, EqualRanges & equal_ranges) const
+{
+    /// Cast indexes column to the real type so that compareAt and getUInt methods can be inlined.
+    const IndexColumn * real_indexes = assert_cast<const IndexColumn *>(&getIndexes());
+
+    IColumn::Permutation dict_perm;
+    getDictionary().getNestedColumn()->getPermutation(direction, stability, 0, nan_direction_hint, dict_perm);
+
+    PaddedPODArray<UInt64> position_by_index(dict_perm.size());
+    for (size_t i = 0; i < dict_perm.size(); ++i)
+        position_by_index[dict_perm[i]] = i;
```

Comment:
not sure how smart is compiler but maybe create `dict_perm` and `position_by_index` in `updatePermutation` so same code is not generated for each type?

Reply:
moved it to `updatePermutation` as you suggested